File: src/main/java/club/sk1er/mods/levelhead/display/DisplayManager.java
package club.sk1er.mods.levelhead.display;

import club.sk1er.mods.levelhead.config.MasterConfig;
import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import gg.essential.api.EssentialAPI;
import gg.essential.api.utils.JsonHolder;
import org.apache.commons.io.FileUtils;

import java.io.File;
import java.io.IOException;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.List;

public class DisplayManager {

    private final Gson GSON = new Gson();
    private final List<AboveHeadDisplay> aboveHead = new ArrayList<>();
    private LevelheadDisplay chat = null;
    private TabDisplay tab = null;
    private MasterConfig config = new MasterConfig();
    private final File file;

    public DisplayManager(JsonHolder source, File file) {
        if (source == null)
            source = new JsonHolder();
        this.file = file;
        if (source.has("master")) {
            try {
                this.config = GSON.fromJson(source.optJSONObject("master").getObject(), MasterConfig.class);
            } catch (Exception ignored) {

            }
        }
        if (config == null) {
            this.config = new MasterConfig();
            EssentialAPI.getMinecraftUtil().sendMessage("Could not load previous settings! If this is your first time running the mod, nothing is wrong.˚");
        }
        for (JsonElement head : source.optJSONArray("head")) {
            try {
                aboveHead.add(new AboveHeadDisplay(GSON.fromJson(head.getAsJsonObject(), DisplayConfig.class)));
            } catch (Exception ignored) {

            }
        }
        if (source.has("chat")) {
            try {
                this.chat = new ChatDisplay(GSON.fromJson(source.optJSONObject("chat").getObject(), DisplayConfig.class));
            } catch (Exception ignored) {

            }
        }
        if (source.has("tab")) {
            try {
                this.tab = new TabDisplay(GSON.fromJson(source.optJSONObject("tab").getObject(), DisplayConfig.class));
            } catch (Exception ignored) {

            }
        }
        Runtime.getRuntime().addShutdownHook(new Thread(this::save));

        if (aboveHead.isEmpty()) {
            aboveHead.add(new AboveHeadDisplay(new DisplayConfig()));
        }

        if (tab == null) {
            DisplayConfig config = new DisplayConfig();
            config.setType("QUESTS");
            tab = new TabDisplay(config);
        }

        adjustIndexes();

        if (chat == null) {
            DisplayConfig config = new DisplayConfig();
            config.setType("GUILD_NAME");
            chat = new ChatDisplay(config);
        }


    }

    public void adjustIndexes() {
        for (int i = 0; i < aboveHead.size(); i++) {
            aboveHead.get(i).setBottomValue(i == 0);
            aboveHead.get(i).setIndex(i);
        }
    }

    public List<AboveHeadDisplay> getAboveHead() {
        return aboveHead;
    }

    public LevelheadDisplay getChat() {
        return chat;
    }

    public LevelheadDisplay getTab() {
        return tab;
    }

    public MasterConfig getMasterConfig() {
        return config;
    }

    public void tick() {
        if (!config.isEnabled()) {
            return;
        }

        for (AboveHeadDisplay aboveHeadDisplay : aboveHead) {
            aboveHeadDisplay.tick();
        }

        if (tab != null)
            tab.tick();
        if (chat != null)
            chat.tick();
    }

    public void checkCacheSizes() {
        for (AboveHeadDisplay aboveHeadDisplay : aboveHead) {
            aboveHeadDisplay.checkCacheSize();
        }

        if (tab != null) {
            tab.checkCacheSize();
        }
        if (chat != null) {
            chat.checkCacheSize();
        }
    }

    public void save() {
        JsonHolder jsonHolder = new JsonHolder().put("master", new JsonHolder(GSON.toJson(config)));
        if (tab != null) {
            jsonHolder.put("tab", new JsonHolder(GSON.toJson(tab.getConfig())));
        }

        if (chat != null) {
            jsonHolder.put("chat", new JsonHolder(GSON.toJson(chat.getConfig())));
        }

        JsonArray head = new JsonArray();

        for (AboveHeadDisplay aboveHeadDisplay : this.aboveHead) {
            head.add(new JsonHolder(GSON.toJson(aboveHeadDisplay.getConfig())).getObject());
        }

        jsonHolder.put("head", head);

        try {
            FileUtils.writeStringToFile(this.file, jsonHolder.toString(), Charset.defaultCharset());
        } catch (IOException e) {
            e.printStackTrace();
        }

    }

    public void clearCache() {
        for (AboveHeadDisplay aboveHeadDisplay : this.aboveHead) {
            aboveHeadDisplay.cache.clear();
            aboveHeadDisplay.trueValueCache.clear();
        }

        if (tab != null) {
            tab.cache.clear();
            tab.trueValueCache.clear();
        }

        if (chat != null) {
            chat.cache.clear();
            chat.trueValueCache.clear();
        }
    }
}

File: src/main/kotlin/club/sk1er/mods/levelhead/core/DisplayManager.kt
package club.sk1er.mods.levelhead.core

import club.sk1er.mods.levelhead.Levelhead
import club.sk1er.mods.levelhead.Levelhead.gson
import club.sk1er.mods.levelhead.Levelhead.jsonParser
import club.sk1er.mods.levelhead.config.DisplayConfig
import club.sk1er.mods.levelhead.config.MasterConfig
import club.sk1er.mods.levelhead.display.AboveHeadDisplay
import club.sk1er.mods.levelhead.display.ChatDisplay
import club.sk1er.mods.levelhead.display.LevelheadDisplay
import club.sk1er.mods.levelhead.display.TabDisplay
import com.google.gson.JsonArray
import com.google.gson.JsonObject
import gg.essential.universal.UMinecraft
import gg.essential.universal.wrappers.UPlayer
import net.minecraft.entity.player.EntityPlayer
import org.apache.commons.io.FileUtils
import java.io.File
import java.io.IOException
import java.nio.charset.StandardCharsets

class DisplayManager(val file: File) {

    var config = MasterConfig()
    val aboveHead: MutableList<AboveHeadDisplay> = ArrayList()
    lateinit var chat: ChatDisplay
        private set
    lateinit var tab: TabDisplay
        private set

    init {
        this.readConfig()
    }

    fun readConfig() {
        try {
            var shouldSaveCopyNow = false
            if (!file.exists()) {
                file.createNewFile()
                shouldSaveCopyNow = true
            }
            val source = jsonParser.parse(FileUtils.readFileToString(file)).runCatching { asJsonObject }.getOrElse { JsonObject() }
            if (source.has("master")) this.config = gson.fromJson(source["master"].asJsonObject, MasterConfig::class.java)

            if (source.has("head")) {
                for (head in source["head"].asJsonArray) {
                    aboveHead.add(AboveHeadDisplay(gson.fromJson(head.asJsonObject, DisplayConfig::class.java)))
                }
            }

            if (aboveHead.isEmpty()) {
                aboveHead.add(AboveHeadDisplay(DisplayConfig()))
            }

            if (source.has("chat"))
                this.chat = ChatDisplay(gson.fromJson(source["chat"].asJsonObject, DisplayConfig::class.java))
            else
                this.chat = ChatDisplay(DisplayConfig().also { it.type = "GUILD_NAME" })

            if (source.has("tab"))
                this.tab = TabDisplay(gson.fromJson(source["tab"].asJsonObject, DisplayConfig::class.java))
            else
                this.tab = TabDisplay(DisplayConfig().also { it.type = "QUESTS" })

            adjustIndices()

            if (shouldSaveCopyNow) saveConfig()

        } catch (e: IOException) {
            Levelhead.logger.error("Failed to initialize display manager.", e)
        }
    }

    fun saveConfig() {
        val jsonObject = JsonObject()

        jsonObject.add("master", gson.toJsonTree(config))
        jsonObject.add("tab", gson.toJsonTree(tab.config))
        jsonObject.add("chat", gson.toJsonTree(chat.config))

        val head = JsonArray()

        this.aboveHead.forEach { display ->
            head.add(gson.toJsonTree(display.config))
        }

        jsonObject.add("head", head)

        try {
            FileUtils.writeStringToFile(file, jsonObject.toString(), StandardCharsets.UTF_8)
        } catch (e: IOException) {
            Levelhead.logger.error("Failed to write to config.", e)
        }
    }

    fun adjustIndices() {
        for (i in aboveHead.indices) {
            aboveHead[i].bottomValue = i == 0
            aboveHead[i].index = i
        }
    }

    fun joinWorld() {
        val displays = mutableListOf(chat, tab).also { it.addAll(aboveHead.filterIndexed{ i, _ -> i <= Levelhead.LevelheadPurchaseStates.aboveHead}) }
        UMinecraft.getWorld()!!.playerEntities
            .map { playerInfo ->
                displays.map {
                Levelhead.LevelheadRequest(playerInfo.uniqueID.trimmed, it,
                    if (it is AboveHeadDisplay) it.bottomValue else false
                )
            } }.flatten().chunked(20).forEach { reqList ->
                Levelhead.fetch(reqList)
            }
    }

    fun playerJoin(player: EntityPlayer) {
        if (player.isNPC) return
        val displays = mutableListOf(chat, tab).also { it.addAll(aboveHead.filterIndexed{ i, _ -> i <= Levelhead.LevelheadPurchaseStates.aboveHead}) }
        displays.filter { !it.cache.containsKey(player.uniqueID) }
            .map { Levelhead.LevelheadRequest(player.uniqueID.trimmed, it,
                if (it is AboveHeadDisplay) it.bottomValue else false
            ) }.ifEmpty { return }.run { Levelhead.fetch(this) }
    }

    fun update() {
        aboveHead.forEachIndexed { i, head ->
            if (i == 0 && Levelhead.LevelheadPurchaseStates.customLevelhead) return@forEachIndexed
            if (i > Levelhead.LevelheadPurchaseStates.aboveHead) return@forEachIndexed
            head.cache.forEach { (_, tag) ->
                if (tag.owner == UPlayer.getUUID() && i == 0 &&
                            Levelhead.LevelheadPurchaseStates.customLevelhead
                        ) return@forEach
                tag.header.run {
                    this.chroma = head.config.headerChroma
                    this.color = head.config.headerColor
                    this.value = "${head.config.headerString}: "
                }
                tag.footer.run {
                    this.chroma = head.config.footerChroma
                    this.color = head.config.footerColor
                }
            }
        }
        chat.run {
            this.cache.forEach { (_, tag) ->
                tag.header.let {
                    it.chroma = this.config.headerChroma
                    it.color = this.config.headerColor
                }
                tag.footer.let {
                    it.chroma = this.config.footerChroma
                    it.color = this.config.footerColor
                }
            }
        }
        tab.run {
            this.cache.forEach { (_, tag) ->
                tag.header.let {
                    it.chroma = this.config.headerChroma
                    it.color = this.config.headerColor
                }
                tag.footer.let {
                    it.chroma = this.config.footerChroma
                    it.color = this.config.footerColor
                }
            }
        }
    }

    fun checkCacheSizes() {
        aboveHead.forEachIndexed { i, head ->
            if (i > Levelhead.LevelheadPurchaseStates.aboveHead) return@forEachIndexed
            head.checkCacheSize()
        }
        chat.checkCacheSize()
        tab.checkCacheSize()
    }

    fun clearCache() {
        aboveHead.forEachIndexed { i, head ->
            if (i > Levelhead.LevelheadPurchaseStates.aboveHead) return@forEachIndexed
            head.cache.clear()
        }
        chat.cache.clear()
        tab.cache.clear()
        joinWorld()
    }
}